### PR TITLE
feat: implement custom callback for static legend

### PR DESCRIPTION
**Pull request by TCL735**
Part of https://github.com/influxdata/ui/issues/1523

Implements a custom callback function that runs on every render of the static legend
- allows the user to get detailed information about the styling of the static legend box
- creates a hook for auto-adjusting the static legend
- separates the styling of static legend and tooltip to allow more specialization between them

Screen Shot 2021-06-14 at 4 26 49 PM

Screen Shot 2021-06-14 at 4 26 36 PM

### CODE REVIEW

**Comment by TCL735** [comment]
Force push to fix merge conflicts.

Ready for review.

**Review by hoorayimhelping — COMMENTED** [review]
a couple of small questions

File: giraffe/src/components/LegendStyles.ts [hunk 1]
```diff
@@ -186,6 +199,46 @@ const legendColumnValueStyle = (
...
+  colors: string[],
+  colorizeRows: boolean,
+  fontBrightColor: string,
+  switchToVertical: boolean
+): React.CSSProperties => {
+  let color = fontBrightColor
+
+  if (colorizeRows && colors) {
+    color = colors[i]
+  }
```
[the remark this answers is not in the document]
Comment (TCL735) [reply giraffe/src/components/LegendStyles.ts:213]: CSS handles this correctly and you get no color. No error is thrown and the app won't break.

File: giraffe/src/components/StaticLegendBox.tsx [hunk 2]
```diff
@@ -62,18 +72,33 @@ export const StaticLegendBox: FunctionComponent<StaticLegendBoxProps> = props =>
...
-    valueColumnKey,
-    columnFormatter
+  const legendData = useMemo(
+    () =>
+      getLegendData(layerConfig.type, spec, valueColumnKey, columnFormatter),
+    [layerConfig.type, spec, valueColumnKey, columnFormatter]
   )
 
+  useEffect(() => {
+    staticLegendOverride.renderEffect({
```
[the remark this answers is not in the document]
Comment (TCL735) [reply giraffe/src/components/StaticLegendBox.tsx:82]: Yes, it is memoized in this way: the `useEffect`'s dependency array will skip the callback if any of the dependencies did not change. Each of the dependencies in the array are memoized.

**Comment by TCL735** [comment]
While working in the ui, i decided to rename `sampleRowTextMetrics` to something more generic and picked `sampleTextMetrics`

**Review by hoorayimhelping — APPROVED** [review]

**Pull request merged**